src/CorePolicy.cc

```diff
@@ -0,0 +1,247 @@
+    Arachne::PerfStats::collectStats(&previousStats);
+
+    // Each loop cycle makes measurements and uses them to evaluate whether
+    // to increment the core count, decrement it, or do nothing.
+    for (Arachne::PerfStats currentStats; ; previousStats = currentStats) {
+        Arachne::sleep(MEASUREMENT_PERIOD);
+        Arachne::PerfStats::collectStats(&currentStats);
+
+        // Take a snapshot of currently active cores before performing
+        // estimation to avoid races between estimation and the fulfillment of
+        // a previous core request.
+        uint32_t curActiveCores =
+          corePolicy->getCoreList(corePolicy->defaultClass)->numFilled;
+
+        // Evaluate idle time precentage multiplied by number of cores to
+        // determine whether we need to decrease the number of cores.
+        uint64_t idleCycles =
+            currentStats.idleCycles - previousStats.idleCycles;
+        uint64_t totalCycles =
+            currentStats.totalCycles - previousStats.totalCycles;
+        uint64_t utilizedCycles = totalCycles - idleCycles;
+        uint64_t totalMeasurementCycles =
+            Cycles::fromNanoseconds(MEASUREMENT_PERIOD);
+        double totalUtilizedCores =
+            static_cast<double>(utilizedCycles) /
+            static_cast<double>(totalMeasurementCycles);
+
+        // Estimate load to determine whether we need to increment the number
+        // of cores.
+        uint64_t weightedLoadedCycles =
+            currentStats.weightedLoadedCycles -
+            previousStats.weightedLoadedCycles;
+        double averageLoadFactor =
+            static_cast<double>(weightedLoadedCycles) /
+            static_cast<double>(totalCycles);
+        if (curActiveCores < Arachne::maxNumCores &&
+                averageLoadFactor > loadFactorThreshold) {
+            // Record our current totalUtilizedCores, so we will only ramp down
+            // if utilization would drop below this level.
+            utilizationThresholds[curActiveCores] = totalUtilizedCores;
+            Arachne::incrementCoreCount();
+            continue;
+        }
+
+        // We should not ramp down if we have high occupancy of slots.
+        double averageNumSlotsUsed = static_cast<double>(
+                currentStats.numThreadsCreated -
+                currentStats.numThreadsFinished) /
+                curActiveCores / Arachne::maxThreadsPerCore;
+
+        // Scale down if the idle time after scale down is greater than the
+        // time at which we scaled up, plus a hysteresis threshold.
+        if (totalUtilizedCores < utilizationThresholds[curActiveCores - 1]
+                - idleCoreFractionHysteresis &&
+                averageNumSlotsUsed < SLOT_OCCUPANCY_THRESHOLD) {
+            Arachne::decrementCoreCount();
+        }
+    }
+}
+
+/*
+ * Block a core in the kernel.
+ *
+ * \param coreId
+ *     The coreId of the core that will block
+ */
+void CoreBlocker::blockCore(int coreId) {

```

Comment:
I'm not sure that "blockCore" is the best name for this; that feels like it describes the implementation more than the abstraction. Let's discuss better possible names.